components/scheduler/executor_executors.go

```diff
@@ -0,0 +1,493 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package scheduler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	schedulepb "go.temporal.io/api/schedule/v1"
+	workflowpb "go.temporal.io/api/workflow/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/server/api/historyservice/v1"
+	schedulespb "go.temporal.io/server/api/schedule/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/server/common/util"
+	"go.temporal.io/server/service/history/hsm"
+	scheduler1 "go.temporal.io/server/service/worker/scheduler"
+	"go.uber.org/fx"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	ExecutorTaskExecutorOptions struct {
+		fx.In
+
+		Config         *Config
+		MetricsHandler metrics.Handler
+		BaseLogger     log.Logger
+		HistoryClient  resource.HistoryClient
+		FrontendClient workflowservice.WorkflowServiceClient
+	}
+
+	executorTaskExecutor struct {
+		ExecutorTaskExecutorOptions
+	}
+
+	rateLimitedDetails struct {
+		// The requested interval to delay processing by rescheduilng.
+		Delay time.Duration
+	}
+)
+
+const (
+	startWorkflowMinDeadline = 5 * time.Second // Lower bound for the deadline in which buffered actions are dropped.
+	startWorkflowMaxDeadline = 1 * time.Hour   // Upper bound for the deadline in which buffered actions are dropped.
+
+	// Because the catchup window doesn't apply to a manual start, pick a custom
+	// execution deadline before timing out a start.
+	manualStartExecutionDeadline = startWorkflowMaxDeadline
+
+	// Upper bound on how many times starting an individual buffered action should be retried.
+	ExecutorMaxStartAttempts = 10
+
+	errTypeRetryLimitExceeded = "RetryLimitExceeded"
+	errTypeRateLimited        = "RateLimited"
+	errTypeAlreadyStarted     = "serviceerror.WorkflowExecutionAlreadyStarted"
```

Comment:
Ideally we'd use the standard errors package instead of performing string comparisons.